### DISCUSSION TITLE
AI frequently gives away guaranteed first strike when fielding Fast troops against Average

### DISCUSSION THREAD

**Discussion by Granfalegion** [ihhub/fheroes2#7824] [General]
fHeroes Battle AI is generally good about moving troops who are faster than their opponents to positions that guarantee them the first strike.  That is, a Champion is faster than an Ogre, and knows that he can move to 3 or more hexes away from the Ogre safely.  The Champion will get the first hit because the Ogre (_morale pending_) will not be able to reach the Champion.

For some reason, however, fHeroes AI seems to regularly get this wrong when ground troops of Fast and Average speeds face off against one another.  When two such stacks are deployed opposite one another and no terrain interferes, the troops will reach each other when they both move at full speed right at one another.  In single combat, we observe that the Fast troop moves his _full_ distance toward his opponent during his turn.  The Average troop then spends his turn moving his full distance also, reaching the Fast troop and getting the first blow in combat.

Here's an example of this, as shown in a Hot Seat combat between a stack of Paladins (Fast) vs Water Elementals (Average): https://clips.twitch.tv/GiantCarelessToothGingerPower-GVHwjn0Q8UGtvpUN

The Paladins move into range of the Water Elementals when they don't need to and end up attacking second on the first turn of combat.  This is especially bad for Paladins, as they give up not just first strike, but their bonus attack on the same move.

This is on v1.0.8, I believe.

**Comment by oleg-derevenetz** [comment]
Hi @Granfalegion here is #7827 which is designed to solve similar problems (and some others).

**Comment by Granfalegion** [comment]
Nice!
